// _local/models/content.js
var _ = require('lodash');
var mongoose = require('mongoose');
var filter = require('feathers-query-filters');

var schema = {
  schema: {
    name: {
      type: String,
      required: '{PATH} is required!',
    },
    relPath: {
      type: String,
      required: '{PATH} is required!',
      index: { unique: true },
    },
    type: {
      type: String,
      default: 'User',
      enum: ['news'],
    },
    fields: {
      type: mongoose.Schema.Types.Mixed,
      default: {},
    },
    updated_at: {
      type: Date,
      default: new Date(),
    },
    created_at: {
      type: Date,
      default: new Date(),
    },
  },
  pre:{
    save: [function(next) {
      var that = this;
      that.updated_at = _.isUndefined(that.created_at) ? that._id.getTimestamp() : new Date();
      if (!that.created_at)
        that.created_at = that._id.getTimestamp();
      next();
    }]
  },
  methods: {
  },
  statics: {
  },
  virtuals: {
  },
  indexes: [
  ],
  // Hooks
  before:{
    //all: [authHooks.requireAuth],
    //find: [authHooks.queryWithUserId],
    //get: [authHooks.queryWithUserId],
    // These will be executed in the order listed
    //create: [authHooks.setUserId, hooks.log],
    //update: [authHooks.setUserId],
    //patch: [authHooks.setUserId],
    //remove: [authHooks.setUserId]
  },
  after:{
    //all: [],
    //find: [hooks.log],
    //get: [],
    //create: [],
    //update: [],
    //patch: [],
    //remove: []
  },

  //add custom endpoints here, see https://github.com/feathersjs/feathers/pull/91
  _setup: function(app, path) {
    var self = this;
    var model = this.model

    //Count - mostly taken from find, used to get the total count of a query.
    app.get('/' + path + '/count', function (req, res, next) {
      var params = req.params;
      params.query = params.query || {};
      var filters = filter(params.query);

      var query = self.model.count(params.query);

      if (filters.$select && filters.$select.length) {
        var fields = {};

        _.each(filters.$select, function(key){
          fields[key] = 1;
        });

        query.select(fields);
      }

      if (filters.$sort){
        query.sort(filters.$sort);
      }

      if (filters.$limit){
        query.limit(filters.$limit);
      }

      if (filters.$skip){
        query.skip(filters.$skip);
      }

      if (filters.$populate){
        query.populate(filters.$populate);
      }

      query.exec(function(error, data) {
        if (error) {
          return next(new errors.BadRequest(error));
        }

        res.json({count: data});
      });
    });
  }
};
module.exports = schema;

/*
var schema = mongoose.Schema({
  name: {
    type: String,
    required: '{PATH} is required!',
  },
  relPath: {
    type: String,
    required: '{PATH} is required!',
    index: { unique: true },
  },
  type: {
    type: String,
    default: 'User',
    enum: ['news'],
  },
  fields: {
    type: mongoose.Schema.Types.Mixed,
    default: {},
  },
  updated_at: {
    type: Date,
    default: new Date(),
  },
  created_at: {
    type: Date,
    default: new Date(),
  },
});


schema.pre('save', function(next) {
  var that = this;
  that.updated_at = _.isUndefined(that.created_at) ? that._id.getTimestamp() : new Date();
  if (!that.created_at)
    that.created_at = that._id.getTimestamp();
  next();
});

module.exports = mongoose.model('Content', schema);

*/
